L80-csharp-load-balancing.md

```diff
@@ -0,0 +1,81 @@
+  * A collection of load balancer factories are fetched from the service provider. One is selected using the configured name (if none specified then `pick_first` is used) and it will create load balancer.
+  * Load balancer is initialized with an `IChannelController` instance by the channel. This is used to call back to the channel to create sub-channels and update the picker.
+
+### Connecting
+
+* No network activity until `GrpcChannel.ConnectAsync` is called, or the first RPC is started.
+* On connect, the `ConnectionManager` will subscribe to address resolver and call the resolver's `RefreshAsync` to prompt the resolver to run immediately.
+* Resolver will push addresses to `ConnectionManager` which will then call balancer.
+* Balancer will use `IChannelController` to create, remove or update sub-channels as needed based on the addresses.
+  * A new sub-channel will start connecting when `SubChannel.ConnectAsync` is called. A sub-channel will be shutdown when disposed.
+* Balancer will be notified on changes in sub-channel state. It will create a `SubChannelPicker` based on the current state.
+
+### Picking
+
+* `ConnectionManager.PickAsync` is called to get a ready sub-channel along with the sub-channel's connected address.
+* If a picker hasn't been set yet, or the picker or empty or an error then `PickAsync` will wait for a picker with a ready sub-channel. This depends on whether `WaitForReady` has been set.
+
+## Rationale
+
+* Design shares the most in common with grpc-go. SubChannels have a collection of addresses and are responsible for staying connected and retrying when in TRANSIENT_FAILURE.
+* Goal is to support basic gRPC load balancing using DNS in Kubernetes.
+* Secondary goal is to have a design in place that can be used to implement xDS in the future with minimal changes.
+* All load balancing APIs will be marked as experimental.
+* API design of resolver and load balancer purposefully choses not to use static configuration. This is considered an anti-pattern in .NET. Instead custom resolver and load balancer factories are registered with a DI container which is then hooked up to the gRPC channel using `GrpcChannelOptions.ServiceProvider`.
+
+## Implementation
+
+Implementation is in the grpc-dotnet repo: [PR #1286](https://github.com/grpc/grpc-dotnet/pull/1286).
+
+A diff of changes in public API surface are available [here](https://apiview.dev/Assemblies/Review/2fe5615b41d240f694c5f434c6db3f5b?diffRevisionId=acc91a0dba6841b48c72721e9c2c28b9&doc=False&diffOnly=False).
+
+## Open issues (if applicable)
+
+* There is a race-condition between a picker returning a ready sub-channel and it being used, and the sub-channel's state or address changing. For this reason a picker will return a sub-channel along with its current address when it was marked as ready.
+* DNS support in .NET doesn't support getting service config from DNS. Resolver updating the service config is not part of the current design. Can change if desired.
```

Comment:
I do think that having the resolver return the service config needs to be part of this architecture.  Even if your DNS resolver implementation can't do that, it is something that will definitely be needed by other resolver implementations, including xDS.

Reply:
I agree.

A new service config means the load balancer could change. What happens to subchannels when the load balancer changes? Are they all closed and recreated by the new load balancer?

Reply:
Updated implementation so the load balancer can be changed by the resolver.

At the moment all subchannels are shutdown and recreated if the load balancer changes. I noticed that Java kind of has a pool for this situation. When a subchannel is removed then it is put into the pool with a timer to remove it. A load balancer creating a subchannel asks the pool if it has a subchannel with matching addresses and could reuse a previous subchannel. This will stop the timer.

What Java does could be retrofitted in later if it is needed.